### PR TITLE
Fix module resolution in `include!` macro 

### LINKED ISSUE #7147: `include!` mod declaration not working
## Environment

* **IntelliJ Rust plugin version:** 0.3.145.3797-211
* **Rust toolchain version:** 1.53.0-nightly (5a4ab2645 2021-04-18) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2021.1 (CL-211.6693.114)
* **Operating system:** Linux 5.8.0-49-generic
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description

Included an rs file with module declaration as `pub mod xxx;`, it cannot resolve that module as expected.

## Steps to reproduce

```
.
├── Cargo.toml
├── src
│  └── lib.rs
└── tmp
   ├── lib.rs
   └── zzz.rs
```

`tmp/zzz.rs`:
```rust
pub fn zfn() {}
```
Screenshot from 2021-04-20 02-57-07

`tmp/lib.rs`:
```rust
pub mod zzz;
```
Screenshot from 2021-04-20 02-56-13

`src/lib.rs`:
```rust
include!("../tmp/lib.rs");

pub fn yah() {
    zzz::zfn();
}
```
Screenshot from 2021-04-20 02-57-33

### PR DESCRIPTION
**Pull request by zhongzc**
fix #7147 

changelog: Fix module resolution in `include!` macro

### CODE REVIEW

**Comment by zhongzc** [comment]
cc @Undin

File: src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt [hunk 1]
```diff
@@ -17,6 +17,7 @@ import org.rust.lang.RsConstants
 import org.rust.lang.RsFileType
 import org.rust.lang.core.crate.Crate
 import org.rust.lang.core.macros.RangeMap
+import org.rust.lang.core.macros.includedFrom
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt:20]: Redundant import
```suggestion

```

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt [hunk 2]
```diff
@@ -31,6 +31,7 @@ import org.rust.cargo.util.AutoInjectedCrates.CORE
 import org.rust.cargo.util.AutoInjectedCrates.STD
 import org.rust.ide.injected.isDoctestInjection
 import org.rust.lang.RsConstants
+import org.rust.lang.RsFileType
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt:34]: Redundant import
```suggestion

```

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt [hunk 3]
```diff
@@ -57,6 +58,7 @@ import org.rust.lang.core.types.infer.foldCtConstParameterWith
 import org.rust.lang.core.types.infer.foldTyTypeParameterWith
 import org.rust.lang.core.types.infer.substitute
 import org.rust.lang.core.types.ty.*
+import org.rust.openapiext.findFileByMaybeRelativePath
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt:61]: Redundant import
```suggestion

```

File: src/test/kotlin/org/rust/lang/core/resolve/RsIncludeMacroResolveTest.kt [hunk 4]
```diff
@@ -78,6 +78,340 @@ class RsIncludeMacroResolveTest : RsResolveTestBase() {
...
+            baz::yin::yang();
+                      //^ foo/baz/yin/mod.rs
+        }
+    //- foo/bar.rs
+        pub mod baz;
+    //- foo/baz.rs
+        pub mod yin;
+    //- foo/baz/yin/mod.rs
+        pub fn yang() {}
+    """)
```
Comment (vlad20012) [inline src/test/kotlin/org/rust/lang/core/resolve/RsIncludeMacroResolveTest.kt:131]: Could you, please, add these tests? They work fine, by the way.

```
fun `test resolve function from nested module declared in included file 3`() = checkResolve("""
    //- lib.rs
        include!("foo/bar.rs");
        fn foobar() {
            baz::yin::yang();
                      //^ foo/baz/yin.rs
        }
    //- foo/bar.rs
        pub mod baz {
            pub mod yin;
        }
    //- foo/baz/yin.rs
        pub fn yang() {}
    """)

    fun `test resolve function from nested module declared in included file 4`() = checkResolve("""
    //- lib.rs
        include!("foo/bar.rs");
        fn foobar() {
            baz::yin::yang::quux();
                          //^ foo/baz/yin/yang.rs
        }
    //- foo/bar.rs
        pub mod baz {
            pub mod yin {
                pub mod yang;
            }
        }
    //- foo/baz/yin/yang.rs
        pub fn quux() {}
    """)
```

**Review by vlad20012 — COMMENTED** [review]
Hello! Thanks for the PR! I'm sorry for so long response.

I see you're a first-time contributor. Could you, please, sign our `Contributor License Agreement` as mentioned in [CONTRIBUTING.md](https://github.com/intellij-rust/intellij-rust/blob/master/CONTRIBUTING.md#cla) and add your username to `CONTRIBUTORS.txt`?

Also, could you please squash your commits into one?

**Comment by zhongzc** [comment]
Later I found that it can't pass this test:
```kt
    fun `test resolve nested include macros`() = checkResolve("""
    //- lib.rs
        include!("foo/bar.rs");
        fn foo(f: foobar::FooBar) {}
                           //^ foo/foobaz.rs
    //- foo/bar.rs
        pub mod foobar {
            include!("foobaz.rs");
        }
    //- foo/foobaz.rs
        pub struct FooBar;
    """)
```

and I cannot figure out a satisfactory solution. Besides, extra branches introduced by this patch increase the complexity. Do you think about a better way to solve it? @vlad20012

File: src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt [hunk 5]
```diff
@@ -362,16 +365,15 @@ private class ModCollector(
     /** See also [processModDeclResolveVariants] */
     private fun resolveModDecl(name: String, pathAttribute: String?): RsFile? {
         val (parentDirectory, fileNames) = if (pathAttribute == null) {
-            val parentDirectory = modData.getOwnedDirectory() ?: return null
+            val parentDirectory =  includeMacroParent ?: modData.getOwnedDirectory() ?: return null
```
Comment (dima74) [inline src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt:368]:
```suggestion
            val parentDirectory = includeMacroParent ?: modData.getOwnedDirectory() ?: return null
```

File: src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt [hunk 6]
```diff
@@ -446,17 +448,18 @@ private sealed class ChildMod(val name: String, val hasMacroUse: Boolean) {
...
  * See also: [RsMod.getOwnedDirectory]
  */
-private fun ChildMod.getOwnedDirectory(parentMod: ModData, pathAttribute: String?): VirtualFile? {
+private fun ChildMod.getOwnedDirectory(parentMod: ModData, pathAttribute: String?, includeMacroParent: VirtualFile?): VirtualFile? {
     if (this is ChildMod.File && name == RsConstants.MOD_RS_FILE) return file.virtualFile.parent
 
     val (parentDirectory, path) = if (pathAttribute != null) {
         when {
             this is ChildMod.File -> return file.virtualFile.parent
+            includeMacroParent != null -> includeMacroParent to pathAttribute
```
Comment (dima74) [inline src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt:457]: It is a bit strange that now `ChildMod.getOwnedDirectory` and `RsMod.getOwnedDirectory` have different structure, though previously they mirror each other (have same set of conditions)

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsMod.kt [hunk 7]
```diff
@@ -50,19 +51,23 @@ interface RsMod : RsQualifiedNamedElement, RsItemsOwner, RsVisible {
      */
     @JvmDefault
     fun getOwnedDirectory(createIfNotExists: Boolean = false): PsiDirectory? {
-        if (this is RsFile && name == RsConstants.MOD_RS_FILE || isCrateRoot) return contextualFile.originalFile.parent
+        if (this is RsFile && name == RsConstants.MOD_RS_FILE || isCrateRoot) return containingFile.originalFile.parent
```
Comment (dima74) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsMod.kt:54]: Why it is needed to change `contextualFile` to `containingFile`?

**Comment by vlad20012** [comment]
@zhongzc

> Later I found that it can't pass this test:
> and I couldn't figure out a satisfactory solution. Besides, extra branches introduced by this patch increase the complexity. Do you think about a better way to solve it?

I guess I'll have to go deep into this to answer.

**Comment by Dominaezzz** [comment]
Any progress on this?

**Comment by vlad20012** [comment]
I had an impressive progress on this, but now it looks like I lost the source code I made :(

@Dominaezzz could you please mention a library or open source project suffering from this issue?

**Comment by Dominaezzz** [comment]
Anything using bindgen would suffer from this but the one I'm using is [esp-idf-sys](https://github.com/esp-rs/esp-idf-sys).

**Comment by vlad20012** [comment]
@Dominaezzz Hmm, have you enabled `org.rust.cargo.evaluate.build.scripts` experimental feature? If not, you can enabled it via the corresponding dialog. To open the dialog, just invoke `Help | Find Action` action and type `Experimental features`

**Comment by Dominaezzz** [comment]
Woah, It works! Awesome! So is this PR no longer needed or is this PR about something else?

**Comment by vlad20012** [comment]
This PR is about name resolution of `mod` declarations (`mod foo;`) inside files included by `include!` macros

**Pull request closed without merging**